### PR TITLE
[BPF] support for ipv4 options

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

Although IPv4 options are not widely used, parsing packets with flexible offsets is a prerequisite for IPv6 processing. The code restructuring needed for IPv4 options will be used for IPv6 processing.

refs https://github.com/projectcalico/calico/issues/4736

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
ebpf: handles IPv4 packets with IP options. XDP only does so in kernels 5.18+ it drops them in older kernels.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf-gpl/tc.c [hunk 1]
```diff
@@ -557,622 +569,843 @@ static CALI_BPF_INLINE void calico_tc_process_ct_lookup(struct cali_tc_ctx *ctx)
...
+
+deny:
+	{
+		struct fwd fwd = {
+			.res = TC_ACT_SHOT,
+			.reason = ctx->fwd.reason,
+		};
+		return fwd;
+	}
+}
```
Comment (tomastigera) [inline felix/bpf-gpl/tc.c:1058]: This essentially splits the original code in do_nat() and post_nat() as we want to reuse the same code in two programs.

File: felix/bpf-gpl/list-ut-objs [hunk 2]
```diff
@@ -11,7 +11,7 @@
 
 emit_filename() {
   echo "bin/test_${from_or_to}_${ep_type}_fib_${log_level}${dsr}.o"
-  echo "bin/test_${from_or_to}_${ep_type}_fib_${log_level}${dsr}_v6.o"
+#  echo "bin/test_${from_or_to}_${ep_type}_fib_${log_level}${dsr}_v6.o"
```
Comment (fasaxc) [inline felix/bpf-gpl/list-ut-objs:14]: WIP?

File: felix/bpf-gpl/xdp.c [hunk 3]
```diff
@@ -61,6 +61,7 @@ int calico_xdp_main(struct xdp_md *xdp)
 		return XDP_DROP;
 	}
 	__builtin_memset(ctx->state, 0, sizeof(*ctx->state));
+	ctx->scratch = (void *)(ctx->xdp_globals + 1);
```
Comment (fasaxc) [inline felix/bpf-gpl/xdp.c:64]: What's going on here?

**Review by fasaxc — COMMENTED** [review]
Looked at everything apart from the extensive changes to tc.c.  Could do with some pointers on what the high-level changes are there.

File: felix/bpf-gpl/xdp.c (hunk 3 above)
Comment (tomastigera) [reply felix/bpf-gpl/xdp.c:64]: although not used in XDP we need to have the pointer set, so pointing behind the globals as there is enough space.

**Comment by tomastigera** [comment]
Rebased and pushed to make sure we use the latest code even though is would still merge cleanly.

**Review by fasaxc — APPROVED** [review]

**Pull request merged**